### ISSUE TITLE
Add ability to re-run single jobs

### ISSUE THREAD

**Issue by tvdeyen** [actions/runner#432]
Please add the ability to re-run single jobs of a workflow. This is such a basic feature.

![alchemy_cms@5d8ae3b 2020-04-17 09-22-32](https://user-images.githubusercontent.com/42868/79544647-a24f3500-808f-11ea-8055-9852b9935944.png)

Please keep the environment in mind while prioritizing features.

🙏

**Comment by TingluoHuang** [comment]
@chrispat from the product team for this feedback. 😄

**Comment by fuesec** [comment]
would love to see this feature

**Comment by schw4rzlicht** [comment]
Any updates on this? In a parallelized Workflow, we always spend 100 build minutes even if one of the 1-minute-jobs fails.

**Comment by rathboma** [comment]
:100: this is a much needed addition.

I maintain [Beekeeper Studio](https://github.com/beekeeper-studio/beekeeper-studio), and random timeouts cause 1/5 jobs to fail fairly regularly. Being able to re-run only the jobs that failed would save us so much time.

**Comment by Ferroin** [comment]
This is a crucial feature for anyone doing multi-arch package builds and deployments with tooling that cannot build multiple architectures in parallel. WIthout this, much more complicated workflows are required to ensure packages don’t get deployed multiple times just because one of the jobs failed.

**Comment by sindrijo** [comment]
This is a much needed feature for my team, which is building a CI pipeline for build/testing/deploying multiple packages for multiple platforms, not having this feature seems very wasteful in both time and electrons.

**Comment by prettycoder** [comment]
have to chime in. My matrix generates 61 jobs and one of them usually fails because of where it collects data from. The next time I re-run 61 jobs another one fails ...

**Comment by tbarbugli** [comment]
Our team is currently exploding 1 workflow into many jobs, this is a terrible hack to get retries to work because you cannot make much sense of the status of a branch/commit anymore (the Actions UI does not do any grouping at that point)

**Comment by OmgImAlexis** [comment]
It's been a year, @TingluoHuang is there any status update on this?

Has it been at least considered?

**Comment by maragunde93** [comment]
Is there any updated on this? we have migrated from GitLab CI to GitHub actions and I am already regretting about this, we have a pipeline that deploys infrastructure and takes around 1 hour, if somethings fails retrying the whole pipeline is a big lost.

**Comment by johanhelsing-attensi** [comment]
I'm honestly quite disappointed that this doesn't seem to have been prioritized, and no response as to when/if it can be expected (just that it was "[definitely on the backlog](https://github.community/t/re-run-jobs/16145/2?u=johanhelsing-attensi)" back in 2019).

In order to provide some substance, and not just spam everyone with complaints and a +1, here's a summary about what I found with respect to workarounds:

https://github.community/t/re-run-jobs/16145/11
> I think I found a workaround here. I have split my matrix build to multiple yaml files, each one with a different name but the same trigger. Each of them contain only a single run. As it looks, this enables me to re-run jobs individually. I do so by selecting “re-run all jobs”, where “all” is now always exactly one.
>
> The price you need to pay is that you have some code duplication and cannot use the matrix feature. For me, this is acceptable as the CI is actually done by a script and shared code outside of that script is relatively minimal.

Unfortunately, this is a bit clunky to use, as at least last I checked neither includes nor yaml anchors are supported in workflows, so code reuse and maintainability across projects will be a pain. Also, I don't really understand how I could make rules like "deploy to staging once *all* builds pass".

The matrix feature is also really nice, and I'd hate to lose it.

There is [another interesting hack](https://github.community/t/ability-to-rerun-just-a-single-job-in-a-workflow/17234/131), which stores the last run status in cache:

How they used it:

```c#
    - name: Set default run status
      run: echo "::set-output name=last_run_status::default" > last_run_status

    - name: Restore last run status
      id: last_run
      uses: actions/cache@v2
      with:
        path: |
          last_run_status
        key: ${{ github.run_id }}-${{ matrix.os }}-${{ matrix.node-version }}-${{ matrix.webpack }}-${{ steps.date.outputs.date }}
        restore-keys: |
          ${{ github.run_id }}-${{ matrix.os }}-${{ matrix.node-version }}-${{ matrix.webpack }}-
    - name: Set last run status
      id: last_run_status
      run: cat last_run_status

    - name: Checkout ref
      uses: actions/checkout@v2
      with:
        ref: ${{ github.event.workflow_dispatch.ref }}

    - name: Use Node.js ${{ matrix.node-version }}
      if: steps.last_run_status.outputs.last_run_status != 'success'
      uses: actions/setup-node@v1
      with:
        node-version: ${{ matrix.node-version }}
```

**Comment by btjones-me** [comment]
+1 we would be grateful for this feature.

**Comment by gergo-papp** [comment]
+1 We are evaluating different CI providers right now (after potentially migrating from Travis) and I'm sure this is a really important feature for many other developers as well

**Comment by Sarga** [comment]
+1 we need this feature.

**Comment by rr-nick-tan** [comment]
looking for this feature too, otherwise, have to split the workflow into multiple ones

**Comment by marcelwa** [comment]
+1 please safe the planet!

**Comment by domdfcoding** [comment]
To avoid spamming everyone with notifications please use GitHub's [reaction buttons](https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) instead of commenting "+1 we want this". Thanks 😃

**Comment by tonyvince** [comment]
@chrispat is this something that can be expected soon?

**Comment by hsalem7** [comment]
I had to split the workflows so I don't have to run a very time-consuming successful job every time another simple job failed.
I have the instructions here if someone really needs this simple nice workaround.
https://hsalem.com/posts/ability-to-rerun-single-jobs-in-github-actions.html

**Comment by iammohitsakhuja** [comment]
Thanks @hsalem7 for your intuitive solution. Is there a way that you might be aware of which allows to pass the output of a job from first workflow to a job of the second workflow?

**Comment by netusco** [comment]
+1 most needed!

**Comment by mlassnig** [comment]
+1

**Comment by jhpratt** [comment]
Everyone: **Please stop responding with "+1" or similar.** You can :+1: the issue as a whole, which serves as an indicator that it's desired. By having numerous people respond with useless crap, you're spamming people's inboxes.

**Comment by barakbd** [comment]
I will just add that both CircleCI and Harness have this ability, and from experience it has proven to be very useful!

**Comment by roberto-vasquez-enermarket** [comment]
+1 This is wasting serious amount of resources. Please add a rerun button to single jobs.

**Comment by bradymholt** [comment]
It appears there is movement on this.  [This tweet](https://twitter.com/kiwiupover/status/1413336136064274436) shows a new "Re-run failed  jobs" item which appears to not be available to everyone but it seems this ability is coming soon.

**Comment by tvdeyen** [comment]
Not for us

https://github.com/AlchemyCMS/alchemy_cms/pull/2047/checks

![Screenshot from 2021-07-14 15-05-40](https://user-images.githubusercontent.com/42868/125627087-1fb0cff1-f558-4490-9732-24fbfd7bbb49.png)

**Comment by bradymholt** [comment]
@tvdeyen - Right...it's not fully rolled out yet.

**Comment by hcguersoy** [comment]
Unfortunately a 're-run failed jobs' feature is only the half of the story and IMHO not enough then you think about _Continuous Deployment_.
Let's take a pipeline there you deploy into multiple stages, each in its own job. 
And you are forced to rollback one stage to a earlier version by triggering only this job in this older pipeline run.

**Comment by master-of-null** [comment]
Please!

**Comment by nishwalshetty** [comment]
Hey All,
when will this feature be rolled out ?

![image](https://user-images.githubusercontent.com/48088420/129680334-c3157092-7eb8-4da8-8eb2-3fe23fc31cc7.png)

**Comment by ashkann** [comment]
Out of curiosity and 100% honest question so maybe we can all learn something here:
Considering the overwhelming amount of +1s and "me too"s about this and the obvious matter of fact that almost any user (paid or otherwise) requires this, how come this hasn't happened yet (512 days now)?

**Comment by Otterian** [comment]
+1, this should be prioritized
- It saves time
- It saves resources
- It's common CI/CD functionality to be allowed to rerun a single build.

**Comment by pat-s** [comment]
Could this issue be locked?

All people subscribed here get a notification for every comment (like this one). 
Rather than commenting you should add your vote in the original post and wait until this feature is available.

**Comment by XVilka** [comment]
It is quite annoying especially in the case of the Windows CI environments. Many projects use Chocolatey to install missing tools and their servers experience network problems from time to time. It happens also on AppVeyor but you can restart only failed jobs there, unlike GitHub Actions. It's especially problematic for the release preparations if you have tons of different jobs with different tests and environments, sometimes they can take hours to finish the whole pipeline. Ability to rerun just one job and its dependents is crucial in my opinion.

**Comment by ppsdatta** [comment]
+1 
This is a much needed feature.

**Comment by OmgImAlexis** [comment]
@pat-s let's not lock this feature still hasn't been added. The "re-run failed jobs" isn't even close to that.

**Comment by pat-s** [comment]
Locking!= closing.

**Comment by Ferroin** [comment]
As much as the ‘me too’ comments are annoying, they would probably be far less prevalent if there was any indication at all here of the status from GitHub themselves. The essentially total silence from them since the original comment about forwarding it to the product team makes it look like they not only are not working on it but also don’t even care, which indirectly encourages people who are used to more traditional forums where the only option to raise awareness is to comment to do so here to try and raise awareness.

@TingluoHuang any chance we could get an actual update on whether this is being worked on or not and if there are any estimates for when it might become generally available? Sorry to ping you directly, but you seem to be the only GitHub employee who is paying any attention at all to this issue and I don’t know of any organizational group I could @mention regarding this.

**Comment by tvdeyen** [comment]
Maybe it's time to spread this to a wider audience? https://twitter.com/tvdeyen/status/1443273117019906049?s=21

**Comment by dinvlad** [comment]
Let's be nice to each other :-) But yes, whole-heartedly agree that it would be great to get some status indication here..

**Comment by pimjaiw** [comment]
I need this feature please.

**Comment by makar2** [comment]
looking forward to this one

**Comment by MartinLesterSynamedia** [comment]
I have solved this using the [dispatch-action](https://github.com/mvasigh/dispatch-action).

I created 1 top level job which calls the dispatch action for each stage. Each stage of my job is split (as many here have already) into individual yaml files and each has a `repository_dispatch:` with a unique sensible name e.g. os_build, os_test, os_push. The top level has all the `on:` events cron, push, etc but if any individual job fails you can just rerun that section using it's manual run option.

It wasn't relevant for me but I see no reason you can't use a matrix at the top level and pass the specific data using the `message` field. You may also benefit from [composite actions](https://docs.github.com/en/actions/creating-actions/creating-a-composite-action) to reduce repetition.

I find it quite a nice way to work and I am not sure I'd change it even if this was implemented as I can create each stage totally independently then integrate it when I am ready.

However like many, many others I came here because sometimes it is just easier, better, cleaner, etc to have a single file with multiple jobs running and I'd like the option to just rerun the set I want and not always rerun all.

**Comment by leon0399** [comment]
It would be really helpful not only for us, GitHub users, but also will reduce load on actions, in cases like this, when actions fail with no reason at all, and to restart this single on we have to restart EVERYTHING, knowing for sure, that there is no error

https://github.com/PHP-Open-Source-Saver/jwt-auth/runs/4641817764

**Comment by XVilka** [comment]
We experience all kinds of network failures (download, git clone) in GHA today, not sure why. Tried to restart all jobs multiple times but some of them fail randomly. A perfect case for this feature, indeed.

**Comment by TingluoHuang** [comment]
@XVilka there was an incident with Actions today about the Hosted runner getting random TSL handshake timeout when making a request to github.com https://www.githubstatus.com/incidents/gh0vvxtlj5v7

**Comment by briantist** [comment]
https://github.community/t/read-timeouts-and-errors-from-github-urls/219615

GitHub didn't acknowledge it as an incident until **at least 30 hours** after it started too. I re-ran one of my workflow runs (and **all 33 of its jobs**), **42 times** before it passed. Since failures were intermittent, and happened early in the process, re-running of a single job would have allowed me to get a passing run in less than an hour, instead of having _an entire day wasted_ re-running everything dozens of times.

**Comment by briantist** [comment]
> It is coming: [github/roadmap#271](https://github.com/github/roadmap/issues/271)
> 
> > GitHub Actions will let you re-run all failed jobs, any individual job, and let you see the results of previous runs.

This was originally in 2021 Q4, however it seems to have slipped and is now marked for 2022 Q1. I really hope it's early Q1.

**Comment by asteigman** [comment]
another +1 for this functionality!

**Comment by jamietdavidson** [comment]
I too switched from GitLab over to GitHub and am experiencing frustration with how much more cumbersome CI/CD has been.  GitLab was so simple and flexible.

Really hoping this comes out on time this year! Please guys, make this happen :).

**Comment by euqen** [comment]
+1

**Comment by Omzig** [comment]
Did you know that *.visualstudio.com can do this in devops?

**Comment by dylanbhughes** [comment]
Hope to see it this quarter 🙏

**Comment by ethomson** [comment]
Hello everyone!  I strongly agree that this is a thing we need - and in fact [this is a thing that we're working on](https://github.com/github/roadmap/issues/271).  However, this is a part of Actions itself, it's not a part of the runner application (meaning: the software that's in this repository).

In order to keep things tidy for the runner team - the developers who are working on this application - I'm going to close this issue where it will stay off of their bug list.

This is being tracked in [our feedback repository](https://github.com/github/feedback/discussions/4278) which is where you can request features in GitHub Actions.  Thanks for all the feedback, everyone, and I hope to see you [in our feedback repo](https://github.com/github/feedback/discussions/4278).

**Comment by bartlettroscoe** [comment]
NOTE: Avoiding rerunning jobs that have already passed is more than just saving computing cycles.  It also is critical to avoid the ***cumulative probability of failures*** in the different jobs that can significantly increase the number of testing iterations needed to get all passing jobs.  For example, if you have a GitHub Actions setup with seven independent jobs that run to test a PR, if there is a 20% chance of a random failure in any one of the seven PR builds, then the chance of having at least one of the PR builds having a failure jumps to 1 - (1 - 0.2)^7 = 0.79 or 80%!    And if any job fails, you have to rerun all of the jobs and the probability of failure the next time is still 80% and so on.  The result is that it can take many PR testing iterations to get all of the jobs to pass.

This occurs relatively frequently, for example,  in the [Trilinos](https://github.com/trilinos/Trilinos) PR testing system (which currently uses a custom PR testing system which suffers from the same inability to rerun individual jobs and where each job has a non-trivial random probability of failure).

What this means is that if you can't rerun single jobs that fail, then you just can't effectively scale to a large number of testing jobs.   As another example, if you have 100 GitHub Actions jobs with just 1% chance of experiencing a failure (which is about the frequency of just being able to fetch dependencies in a GitHub Actions job), then the ***cumulative probability of failure*** across these 100 jobs is 1 - (1-0.01)^100 = 0.63 or 63%!  But if you can rerun individual jobs, the number of GitHub Actions jobs needed to pass goes way down and getting a set of passing jobs becomes much more probable after the first GitHub Actions jobs run that has a 63% cumulative chance of failure.  If there is just a single job that failed in the first running of all of the GHA jobs (due to a random failure), then the rerunning of that one job would have just a 1% chance of failing or a 99% of passing.  That reduces wasted computing resources and speeds up the testing cycle wall-clock time.

This is a big deal for projects that need many testing jobs and have a higher probability of failure in any individual job.

**Comment by piotrekkr** [comment]
Seems like it is live now and we can rerun single jobs. I'm really grateful for devs for implementing this :pray: And now some tiny rant :) 

It kinda broken our setup when using job matrix and Cypress parallel tests... 

### Here is how it worked and why it is broken now

1. on "setup" job we generated unique ID for cypress tests run
2. next we used job matrix to generate three workers that were running Cypress tests in parallel using generated ID
3. when some jobs in matrix fail we rerun full workflow which generated new ID and run whole matrix again

Why it does not work with rerunning only one matrix job? Because unique ID id the same and Cypress consider this run as finished and do not run tests again. I did not find a way to force running them again with same ID. Whet can be done wit this is:

1. rerun whole workflow again (old way)
2. rerun failed jobs only (will create new workflow run attempt only with failed matrix jobs and with same unique ID)
3. rerun manually all failed jobs one by one (no way to manually rerun whole matrix again)
4. rerun "setup" job will trigger new ID and also will trigger all dependent jobs

First approach is slow since all need to be rerun again. Second approach seems best at first because we could use `UNIQUE_ID-RUN_ATTEMPT` as Cypress ID, but it can be problematic when only 1 of 10 matrix jobs failed and one worker will need to handle all e2e tests again (no parallelization). Third approach is not good either since we cannot select multiple jobs to rerun manually so no parallelization. Last approach is what we use now and it works ok but developers need to remember to rerun this setup job instead of just rerunning failed jobs.

### So to sum up 

Maybe we could add some flag to mark whole matrix as failed when one of jobs inside martix fails? When we rerun all failed jobs it would rerun whole matrix again? Whet do you think? 

Thanks

**Comment by tvdeyen** [comment]
Wow. Finally. **Two years** of wasting precious resources later it finally shipped. Thanks for everyone involved.

<img alt="alchemy_cms@1715663 2022-03-16 11-42-17" src="https://user-images.githubusercontent.com/42868/158572469-a0c87427-8daa-441e-b1cc-4bcfd5dfee45.png">

**Comment by Jolg42** [comment]
Can confirm it's here! 🎊 
<img width="358" alt="Screen Shot 2022-03-16 at 11 22 22" src="https://user-images.githubusercontent.com/1328733/158593441-29256e9a-bbde-4096-aeb6-7c424d73311a.png">
<img width="608" alt="Screen Shot 2022-03-16 at 11 22 58" src="https://user-images.githubusercontent.com/1328733/158593450-1aaaa882-422c-47e1-abbf-72b567da8406.png">

Looks like Santa was early this year 🎅🏼

**Comment by willyt150** [comment]
I saw the [release announcement for supporting re-running single jobs](https://github.blog/2022-03-16-save-time-partial-re-runs-github-actions/), is this being released in phases or something? The GitHub Enterprise repos I'm working on still do not have any ability to re-run individual jobs.

I thought maybe it just wouldn't work with old runs, so I kicked off new ones and still nothing, just the re-run all jobs option.

**Comment by chrispat** [comment]
It is currently available on GitHub.com only and is slated to ship in the next update to GitHub enterprise.  In addition there are still some issues related to reusable workflows that we are ironing out.

**Comment by davegallant** [comment]
> It is currently available on GitHub.com only and is slated to ship in the next update to GitHub enterprise. In addition there are still some issues related to reusable workflows that we are ironing out.

This is amazing work. Not seeing the option to re-run failed jobs for reusable workflows. Wasn't sure if it's because the call to the reusable we're using is dependent upon another job or not.

**Comment by chrispat** [comment]
We have temporarily disabled the feature for any run that references a reusable workflow while we iron out the issues.  We hope to have those resolved towards the end of this week or early next week.

**Comment by debugger24** [comment]
Unable to rerun single job when some jobs are pending review deployment.

<img width="1680" alt="image" src="https://user-images.githubusercontent.com/12707743/160243897-08c5ab99-b9c0-4b70-b173-e8a41f2767b7.png">

**Comment by piotrekkr** [comment]
@debugger24 This is my guess only but this is probably by design. Rerunning any job creates whole new run attempt for whole workflow. All jobs that are not dependent on job you want to rerurn, are "cloned" in new run attempt. But to clone you need a job result first so you need to wait for all jobs to finish.

**Comment by Drowze** [comment]
Found another unexpected behaviour:
- given a job that submits a manual status check (e.g. via API) that has passed
- and given a different job that has failed
- when I retry only failed jobs, the resulting check group will not have the manual check (submitted by the job that has passed on the first try)

This is a problem to us: we have a manual check called "Rubocop" (submitted manually using [reviewdog](https://github.com/reviewdog/reviewdog)) that is required for a pull request to be merged. If we retry the workflow, we have all jobs passing, but the manual check is missing, so a PR can't be merged.

Screenshots detailing the issue:
![Screenshot 2022-03-28 at 14 16 59](https://user-images.githubusercontent.com/9031589/160407174-2799488b-4b86-4bed-a18c-ff08f4834e93.png)
![Screenshot 2022-03-28 at 14 16 44](https://user-images.githubusercontent.com/9031589/160407387-8118506a-9004-4f8c-a142-c5db7ab85523.png)

**Comment by mrmike** [comment]
> We have temporarily disabled the feature for any run that references a reusable workflow while we iron out the issues. We hope to have those resolved towards the end of this week or early next week.

Do you have any public issue opened for this case? I'd like to track progress of this issue

**Comment by hugovk** [comment]
This is now working, thanks!

**Comment by madhavajay** [comment]
Is it possible to re-run a failed job before the others finish? We have quite long running jobs which means the wait to retry a failed test due to some weird external issue is a really long time.
<img width="1010" alt="Screen Shot 2022-05-06 at 9 52 18 am" src="https://user-images.githubusercontent.com/2882739/167044976-2b7999c0-bc7a-440c-86e6-97304b807b0b.png">

**Comment by janpio** [comment]
It is not yet @madhavajay, so I created a feedback discussion to suggest that: https://github.com/orgs/community/discussions/73156 Leave an upvote or reaction over there! (also the 43 other people that upvote the previous comment optimally 😆)

**Comment by abhilash1in** [comment]
I don't see "Re-run failed jobs" option as a dropdown.

I also don't see an option to re-run individual jobs when I hover over them.

Is this a bug or am I doing something wrong?

![bug](https://github.com/actions/runner/assets/9111980/f21d1ddf-5f78-48e0-bd57-ae96008adf34)

**Comment by piotrekkr** [comment]
> I don't see "Re-run failed jobs" option as a dropdown.
> 
> I also don't see an option to re-run individual jobs when I hover over them.
> 
> Is this a bug or am I doing something wrong?

@abhilash1in Are you sure that all jobs inside workflow are finished? If they are not done yet there will be no option to rerun. GitHub requires for full workflow to finish before it can be rerun.

**Comment by abhilash1in** [comment]
> > I don't see "Re-run failed jobs" option as a dropdown.
> > I also don't see an option to re-run individual jobs when I hover over them.
> > Is this a bug or am I doing something wrong?
> 
> @abhilash1in Are you sure that all jobs inside workflow are finished? If they are not done yet there will be no option to rerun. GitHub requires for full workflow to finish before it can be rerun.

Erm, okay all jobs had not finished running when I was looking for the re-run failed jobs button.

But also, that doesn't make sense. If I see failed jobs, I should be able to re-run them individually without having to wait for **all** the jobs to finish.

**Comment by Omzig** [comment]
in azure, i have to wait for the jobs to finish before i can rerun them.............
